pkg/input/container/scanner.go

```diff
@@ -72,29 +72,38 @@ func (c *ContainerInput) run() {
 // scan checks for new containers we're expected to
 // tail, as well as stopped containers
 func (c *ContainerInput) scan(tailFromBegining bool) {
-	containers := c.listContainers()
-	containersIds := make(map[string]bool)
+	runningContainers := c.listContainers()
+	containersToMonitor := make(map[string]bool)
 
-	// monitor new containers
-	for _, source := range c.sources {
-		for _, container := range containers {
+	// list containers to monitor
+	for _, container := range runningContainers {
+		for _, source := range c.sources {
 			if c.sourceShouldMonitorContainer(source, container) {
-				containersIds[container.ID] = true
-				if _, ok := c.tailers[container.ID]; !ok {
-					c.setupTailer(c.cli, container, source, tailFromBegining, c.pp.NextPipelineChan())
-				}
+				containersToMonitor[container.ID] = true
 			}
 		}
 	}
 
 	// stop old containers
 	for containerId, tailer := range c.tailers {
-		if _, ok := containersIds[containerId]; !ok {
+		_, shouldMonitor := containersToMonitor[containerId]
+		if !shouldMonitor || tailer.shouldStop {
 			log.Println("Stop tailing container", containerId[:12])
 			tailer.Stop()
 			delete(c.tailers, containerId)
 		}
 	}
+
+	// monitor new containers
+	for _, container := range runningContainers {
```

Comment:
Could we move this to the other loop to avoid repeating the code ?

Reply:
I'm not sure how this could be done.
We need to do these 3 actions in order:
- list all containers that need to be monitor
- stop monitoring containers that stopped
- start monitoring containers that started

If this doesn't happen in order, we introduce latency / risk missing logs as we'd need to wait for a `scan` to occur before tailing again